toffy/streak_detection.py

```diff
@@ -0,0 +1,429 @@
+    _filter_mask(streak_data=streak_data)
+    return
+
+
+def _streak_correction(streak_data: StreakData, channel: np.ndarray) -> np.ndarray:
+    """Corrects the streaks for the channel argument. Uses masks in the streak_data Dataclass.
+    Performs the correction by averaging the pixels above and below the streak.
+
+    Args:
+        streak_data (StreakData): An instance of the StreakData Dataclass, holds all necessary
+        data for streak correction.
+        channel (np.ndarray): The input channel (Numpy array representation of the tiff file)
+        to be used for streak detection.
+
+    Returns:
+        np.ndarray: The corrected channel.
+    """
+    padded_channel = np.pad(channel.copy(), pad_width=(1, 1), mode="edge")
+    corrected_channel = padded_channel.copy()
+    for region in streak_data.filtered_streak_df.itertuples():
+        corrected_channel[
+            region.max_row, region.min_col : region.max_col
+        ] = _mean_correction(
+            padded_channel,
+            region.min_row,
+            region.max_row,
+            region.min_col,
+            region.max_col,
+        )
+    return util.crop(corrected_channel, crop_width=(1, 1), copy=True)
+
+
+def _mean_correction(
+    channel: np.ndarray, min_row: int, max_row: int, min_col: int, max_col: int
+) -> np.ndarray:
+    """Performs streak-wise correction by: setting the value of each pixel in the streak to the
+    mean of pixel above and below it.
+
+    Args:
+        channel (np.ndarray): The input channel (Numpy array representation of the tiff file) to
+        be used for streak detection.
+        min_row (int): The minimum row index of the streak. The y location where the streak
+        starts.
+        max_row (int): The maximum row index of the streak. The y location where the streak ends.
+        min_col (int): The minimum column index of the streak. The x location where the streak
+        starts.
+        max_col (int): The maximum column index of the streak. The x location where the streak
+        ends.
+
+    Returns:
+        np.ndarray: Returns the corrected streak.
+    """
+    streak_corrected: np.ndarray = np.mean(
+        [
+            channel[min_row, min_col + 1 : max_col + 1],
+            channel[max_row + 1, min_col + 1 : max_col + 1],
+        ],
+        axis=0,
+        dtype=np.int8,
+    )
+
+    return streak_corrected
+
+
+def streak_correction(
+    fov: Union[str, Path],
+    channel: str = "Noodle",
+    mask_data: bool = False,
+) -> Optional[StreakData]:
+    """Takes a `fov` directory and a user specified channel for streak detection. Once all the
+    streaks have been detected on that channel, they are corrected via an averaging method. The
+    function can also return a DataClass containing various binary masks and dataframes which were
+    used for filtering and correction when prompted to.
+
+    Args:
+        fov (Union[str, Path]): A directory containing the fov and all it's channels for
```

Comment:
fov indicates it's the name of the fov. If it's the full path to the fov, the name should be updated accordingly